### DISCUSSION TITLE
CI testing a textual app

### DISCUSSION THREAD

**Discussion by nschloe** [Textualize/textual#166] [Q&A]
For smoke testing, I'd like to start up textual from pytest, e.g.,
```
TiptopApp.run()
```
However, this gives
```
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <_pytest.capture.DontReadFromInput object at 0x7fd0e02a9520>

    def fileno(self) -> int:
>       raise UnsupportedOperation("redirected stdin is pseudofile, has no fileno()")
E       io.UnsupportedOperation: redirected stdin is pseudofile, has no fileno()

.tox/py3/lib/python3.9/site-packages/_pytest/capture.py:229: UnsupportedOperation
```
I'd also need a way for shutting down the app after N seconds or so.

**Comment by 1j01** [comment]
Instead of `App.run`, use [`App.run_test`](https://textual.textualize.io/api/app/#textual.app.App.run_test), which will run the app in headless mode, and thus avoid accessing stdin/stdout.
You can use `run_test`'s context manager (`with` syntax support) to get a `Pilot` object that can be used to interact with the app or wait before exiting. The app is exited automatically when the `with` scope closes.
```py
from your_app.main import YourApp

async def test_app_can_run():
    app = YourApp()
    async with app.run_test() as pilot:
        await pilot.pause(3)
```
To use async tests, you'll need to install `pytest-asyncio` and enable it in `pyproject.toml` like so:
```toml
[tool.pytest.ini_options]
asyncio_mode = "auto"
```
(or it can be enabled in [`pytest.ini`](https://pytest-asyncio.readthedocs.io/en/latest/reference/configuration.html) or with a decorator [`@pytest.mark.asyncio`](https://pytest-asyncio.readthedocs.io/en/latest/reference/markers.html))

----

I also tried using `App.run` with `headless=True`, but it failed to detect errors:
```py
from textual.pilot import Pilot
from your_app.main import YourApp

def test_app_can_run_sync():
    # DON'T USE; this fails to catch errors; I don't know why.
    app = YourApp()
    async def auto_pilot(pilot: Pilot[object]):
        await pilot.pause(3)
        app.exit()  # not needed with run_test
    app.run(headless=True, size=(80, 24), auto_pilot=auto_pilot)
```
